File: src/test/java/no/novari/configuration/validation/constraints/AtLeastOneChildValidatorTest.java
package no.novari.configuration.validation.constraints;

import no.novari.configuration.model.configuration.dtos.ObjectMappingDto;
import no.novari.configuration.model.configuration.dtos.ValueMappingDto;
import no.novari.configuration.validation.constraints.AtLeastOneChildValidator;
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.mockito.Mockito.mock;

class AtLeastOneChildValidatorTest {

    private AtLeastOneChildValidator atLeastOneChildValidator;

    @BeforeEach
    void setup() {
        atLeastOneChildValidator = new AtLeastOneChildValidator();
    }

    @Test
    void shouldReturnTrueWhenObjectMappingHasChildren() {
        ObjectMappingDto objectMappingDto = ObjectMappingDto.builder().build();
        objectMappingDto.getValueMappingPerKey().put("valueMappingKey", mock(ValueMappingDto.class));

        boolean valid = atLeastOneChildValidator.isValid(
                objectMappingDto,
                mock(HibernateConstraintValidatorContext.class)
        );

        assertTrue(valid);
    }

    @Test
    void shouldReturnFalseWhenObjectMappingHasNoChildren() {
        boolean valid = atLeastOneChildValidator.isValid(
                ObjectMappingDto.builder().build(),
                mock(HibernateConstraintValidatorContext.class)
        );

        assertFalse(valid);
    }
}

File: src/test/kotlin/no/novari/flyt/configuration/validation/constraints/AtLeastOneChildValidatorTest.kt
package no.novari.flyt.configuration.validation.constraints

import no.novari.flyt.configuration.model.configuration.dtos.ObjectMappingDto
import no.novari.flyt.configuration.model.configuration.dtos.ValueMappingDto
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock

class AtLeastOneChildValidatorTest {
    private lateinit var atLeastOneChildValidator: AtLeastOneChildValidator

    @BeforeEach
    fun setUp() {
        atLeastOneChildValidator = AtLeastOneChildValidator()
    }

    @Test
    fun `accepts an object mapping that has children`() {
        val objectMappingDto = ObjectMappingDto.builder().build()
        objectMappingDto.valueMappingPerKey["valueMappingKey"] = mock<ValueMappingDto>()

        val valid =
            atLeastOneChildValidator.isValid(
                objectMappingDto,
                mock<HibernateConstraintValidatorContext>(),
            )

        assertTrue(valid)
    }

    @Test
    fun `rejects an object mapping without children`() {
        val valid =
            atLeastOneChildValidator.isValid(
                ObjectMappingDto.builder().build(),
                mock<HibernateConstraintValidatorContext>(),
            )

        assertFalse(valid)
    }
}
